icu4c/source/common/cmemory.h

```diff
@@ -728,6 +735,28 @@ class MemoryPool : public UMemory {
         return fPool[fCount++] = new T(std::forward<Args>(args)...);
     }
 
+    /**
+     * Takes ownership of ptr, adding it to the pool.
+     *
+     * @param ptr Pointer to take ownership of.
+     * @param status Is set to U_MEMORY_ALLOCATION_ERROR if allocation of the
+     * needed pool capacity fails.
+     */
+    template <typename... Args>
+    void adopt(T *ptr, UErrorCode &status) {
```

Comment:
What is the use-case for this method? Is there a use-case where `MemoryPool` is a suitable data structure to use but the `create()` method isn't possible to use?

The reason for why `MemoryPool` currently has a `create()` method but no `adopt()` method is that this makes memory management both simple and unambiguous. When the `MemoryPool` object goes out-of-scope, it deletes all objects that it has created and that's all there is to it.

This proposed `adopt()` method does instead require rather intricate error handling for memory allocation errors:

```
MemoryPool<MyType> pool;

MyType* o1 = new MyType(1, 2, 3);
if (o1 != nullptr) {
  ErrorCode status;
  pool.adopt(o1, status);
  if (status == U_MEMORY_ALLOCATION_ERROR) {
    delete o1;
    o1 = nullptr;
  }
}
```

Compare that to the example in the class comment above that calls the `create()` method.

This looks discouragingly error-prone to me so I wonder what it might be that otherwise isn't possible that this would make possible.

Reply:
My use-case is to have MaybeStackVector hold pointers that were created by a factory method. My usage looks a bit like this:

        fLongNameHandlers.adoptBack(
            LongNameHandler::forMeasureUnit(
                loc, *units[i], MeasureUnit(), width, rules, NULL, status),
            status);

I'm also sad about making MemoryPool more error prone (hence my uncomfortable comment change above). Does it make sense to make it a private method though, for use by MaybeStackVector::adoptBack()?

Reply:
You can't do it like that, your example code snippet will leak the object created by `forMeasureUnit()` if the call to `adoptBack()` fails.

Let's take a step back and look at how you'd like to do memory management in the code you're writing.

If these `LongNameHandler` objects are cheap to copy or move, one option that would simplify memory management considerably would be to let `forMeasureUnit()` return the newly created object by value and then invoke the copy or move constructor through `fLongNameHandlers.emplaceBack()`.

If it isn't possible to make the copy or move constructor cheap enough, an old commonly used pattern is to first create an empty object and then pass a pointer to this object to the factory function that then fills in the data. Like this:

```
LongNameHandler* obj = fLongNameHandlers.emplaceBack();
if (obj != nullptr) {
  LongNameHandler::forMeasureUnit(
              obj, loc, *units[i], MeasureUnit(), width, rules, NULL, status),
}
```

I guess you've seen such code before. It's a bit cumbersome, but it's a very simple way to avoid creating an object that then must be deleted again because it can't be stored anywhere.

And there are other possible solutions, all depending on your bigger picture here.

Reply:
Thanks - I'm tackling a hierarchy of move and copy constructors now (maybe I only need the move constructors).

Reply:
Another pattern might be if `forMeasureUnit` returned the LongNameHandler by value, and then you can move-construct it into the emplaceBack container.

Reply:
No new self-defined constructors necessary, it does seem to work:

        auto *lnh = LongNameHandler::forMeasureUnit(...);
        result->fLongNameHandlers.emplaceBack(*lnh);

And I guess this might do move construction? - But it's still not something I'm confident about:

        result->fLongNameHandlers.emplaceBack(std::move(*lnh));

Reply:
Two problems:

1. The first code example is probably implicitly using the implicitly defined copy constructor.  (This is one of my least favorite things about C++: copy constructor should be explicit, but it's not.)
2. Even if you use `std::move` out of the pointer, like in the second example, you are still doing 2 heap allocations.  Better is to change the signature of `forMeasureUnit` to return by value, such that you can do,

```cpp
result->fLongNameHandlers.emplaceBack(std::move(LongNameHandler::forMeasureUnit(...)));
```

I think the `std::move` above might not be necessary, but it can make it more explicit what you are trying to do.

Reply:
This PR is now so small, I'm wondering if it is still worth submitting/merging separately. :)

Reply:
@sffc I guess I should apply this style principle to LongNameHandler:
https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types
I'll try to do that in my "Glue Code PR" this afternoon (CEST).

Reply:
> The first code example is probably implicitly using the implicitly defined copy constructor.

As of C++11 this copy constructor can be deleted by using the `= deleted` syntax. Even in situations where you want the class to have a copy constructor this can be used during development if you at some point want to verify that the copy constructor isn't called.

Another way to easily verify during development that the intended constructor really is called is to add assertion failures to all other constructors, then a simple run of the unit tests immediately shows exactly which constructor was invoked in any situation where this was unclear.

You can use that to verify that the use of `std::move()` in a situation like this (an object returned by value passed to a move constructor) is a no-op that can (and probably should) be omitted.

Reply:
Copy or move is not possible, because in the process of constructing the instance, other objects are given pointers to it. When it is moved or copied, those pointers fail.

So if we don't add "pointer adoption" to our containers, I'll try adding a fillIn parameter to the factory functions, such that if not null, the fill-in is used, otherwise a new instance is allocated.

Reply:
Refactoring the client code in question to cope with a fillIn: 3a575c9510